File: packages/executable-stories-xunit/ExecutableStories.Xunit/StoryContext.cs
namespace ExecutableStories.Xunit
{
    /// <summary>
    /// Holds per-test BDD story state, managed by <see cref="Story"/> via AsyncLocal.
    /// </summary>
    public class StoryContext(string scenario)
    {
        private static int _orderCounter;
        private sealed class TimerEntry
        {
            public long StartTicks { get; init; }
            public int? StepIndex { get; init; }
            public string? StepId { get; init; }
            public bool Consumed { get; set; }
        }

        public string Scenario { get; } = scenario;
        public List<StoryStep> Steps { get; } = [];
        public List<string> Tags { get; } = [];
        public List<Ticket> Tickets { get; } = [];
        public List<string> Covers { get; } = [];
        public Dictionary<string, object> Meta { get; } = [];
        public List<DocEntry> Docs { get; } = [];
        public int SourceOrder { get; } = Interlocked.Increment(ref _orderCounter) - 1;
        public long StartedTicks { get; } = System.Diagnostics.Stopwatch.GetTimestamp();
        public StoryStep? CurrentStep { get; set; }
        public string? TraceUrlTemplate { get; set; }
        private readonly HashSet<string> _seenPrimary = [];
        private int _stepCounter;
        private readonly List<Dictionary<string, object?>> _attachments = [];
        private readonly Dictionary<int, TimerEntry> _activeTimers = [];
        private int _timerCounter;
        public List<object>? OtelSpans { get; set; }

        public StoryContext(string scenario, params string[] tags) : this(scenario)
        {
            Tags.AddRange(tags);
        }

        /// <summary>
        /// Add a new step and set it as the current step.
        /// </summary>
        public void AddStep(string keyword, string text, DocEntry[]? docs = null)
        {
            // Auto-And: repeated primary keywords render as "And"
            var effective = keyword;
            if (keyword is "Given" or "When" or "Then")
            {
                if (!_seenPrimary.Add(keyword))
                {
                    effective = "And";
                }
            }
            var step = new StoryStep
            {
                Keyword = effective,
                Text = text,
                Id = $"step-{_stepCounter++}"
            };
            Steps.Add(step);
            CurrentStep = step;

            if (docs is { Length: > 0 })
            {
                foreach (DocEntry doc in docs)
                {
                    step.Docs ??= [];
                    step.Docs.Add(doc);
                }
            }
        }

        /// <summary>
        /// Start a high-resolution timer tied to the current step. Returns a token to pass to EndTimer.
        /// </summary>
        public int StartTimer()
        {
            var token = _timerCounter++;
            int? stepIndex = CurrentStep != null ? Steps.Count - 1 : null;
            var stepId = CurrentStep?.Id;
            _activeTimers[token] = new TimerEntry
            {
                StartTicks = System.Diagnostics.Stopwatch.GetTimestamp(),
                StepIndex = stepIndex,
                StepId = stepId,
                Consumed = false
            };
            return token;
        }

        /// <summary>
        /// End the timer and record durationMs on the step that was active when StartTimer() was called.
        /// Double-end is a no-op.
        /// </summary>
        public void EndTimer(int token)
        {
            if (!_activeTimers.TryGetValue(token, out TimerEntry? entry) || entry.Consumed)
            {
                return;
            }

            entry.Consumed = true;

            TimeSpan elapsed = System.Diagnostics.Stopwatch.GetElapsedTime(entry.StartTicks);
            var durationMs = elapsed.TotalMilliseconds;

            StoryStep? step = null;
            if (entry.StepId != null)
            {
                step = Steps.FirstOrDefault(s => s.Id == entry.StepId);
            }
            step ??= entry.StepIndex.HasValue && entry.StepIndex.Value < Steps.Count
                ? Steps[entry.StepIndex.Value]
                : null;

            if (step != null)
            {
                step.DurationMs = durationMs;
            }
        }

        /// <summary>
        /// Attach a doc entry to the current step, or to story-level docs if no step exists yet.
        /// </summary>
        public void AddDoc(DocEntry doc)
        {
            if (doc.Get("children") is IEnumerable<DocEntry> children)
            {
                HashSet<DocEntry> childSet = [.. children];
                _ = Docs.RemoveAll(childSet.Contains);
                foreach (StoryStep step in Steps)
                {
                    if (step.Docs != null)
                    {
                        _ = step.Docs.RemoveAll(childSet.Contains);
                    }
                }
            }

            if (CurrentStep != null)
            {
                CurrentStep.Docs ??= [];
                CurrentStep.Docs.Add(doc);
            }
            else
            {
                Docs.Add(doc);
            }
        }

        /// <summary>
        /// Convert this context to a <see cref="StoryMeta"/> POCO for serialization.
        /// </summary>
        public StoryMeta ToStoryMeta()
        {
            return new StoryMeta
            {
                Scenario = Scenario,
                Steps = Steps.Count > 0 ? new List<StoryStep>(Steps) : null,
                Tags = Tags.Count > 0 ? new List<string>(Tags) : null,
                Tickets = Tickets.Count > 0 ? new List<Ticket>(Tickets) : null,
                Covers = Covers.Count > 0 ? new List<string>(Covers) : null,
                Meta = Meta.Count > 0 ? new Dictionary<string, object>(Meta) : null,
                Docs = Docs.Count > 0 ? new List<DocEntry>(Docs) : null,
                SourceOrder = SourceOrder,
                OtelSpans = OtelSpans
            };
        }

        /// <summary>
        /// Add an attachment to the current story context.
        /// </summary>
        public void AddAttachment(string name, string mediaType, string? path = null, string? body = null, string? encoding = null)
        {
            var a = new Dictionary<string, object?>
            {
                ["name"] = name,
                ["mediaType"] = mediaType,
            };
            if (path != null)
            {
                a["path"] = path;
            }

            if (body != null)
            {
                a["body"] = body;
            }

            if (encoding != null)
            {
                a["encoding"] = encoding;
            }

            if (CurrentStep != null)
            {
                a["stepIndex"] = Steps.IndexOf(CurrentStep);
                a["stepId"] = CurrentStep.Id;
            }
            _attachments.Add(a);
        }

        public List<Dictionary<string, object?>> GetAttachments()
        {
            return _attachments;
        }

        /// <summary>
        /// Reset the global source order counter (useful for tests).
        /// </summary>
        internal static void ResetOrderCounter()
        {
            _ = Interlocked.Exchange(ref _orderCounter, 0);
        }
    }
}

File: packages/executable-stories-junit5/src/main/kotlin/dev/executablestories/junit5/StoryContext.kt
package dev.executablestories.junit5

import java.util.concurrent.atomic.AtomicInteger

class StoryContext(val scenario: String) {

    private class TimerEntry(
        val startNanos: Long,
        val stepIndex: Int?,
        val stepId: String?,
        var consumed: Boolean = false
    )

    val steps: MutableList<StoryStep> = mutableListOf()
    val tags: MutableList<String> = mutableListOf()
    val tickets: MutableList<String> = mutableListOf()
    val meta: MutableMap<String, Any?> = LinkedHashMap()
    val docs: MutableList<DocEntry> = mutableListOf()
    val sourceOrder: Int = ORDER_COUNTER.getAndIncrement()

    var currentStep: StoryStep? = null
        private set

    private val seenPrimaryKeywords: MutableSet<String> = mutableSetOf()
    private var stepCounter = 0
    private val attachments: MutableList<Map<String, Any?>> = mutableListOf()

    private val activeTimers: MutableMap<Int, TimerEntry> = mutableMapOf()
    private var timerCounter = 0

    constructor(scenario: String, vararg tags: String) : this(scenario) {
        this.tags.addAll(tags.toList())
    }

    fun addStep(keyword: String, text: String) {
        addStep(keyword, text, null)
    }

    fun addStep(keyword: String, text: String, vararg docs: DocEntry?) {
        var effectiveKeyword = keyword
        if (isPrimary(keyword)) {
            if (!seenPrimaryKeywords.add(keyword)) {
                effectiveKeyword = "And"
            }
        }
        val step = StoryStep(effectiveKeyword, text)
        step.id = "step-${stepCounter++}"
        steps.add(step)
        currentStep = step
        for (doc in docs) {
            if (doc != null) {
                step.addDoc(doc)
            }
        }
    }

    fun addDoc(doc: DocEntry) {
        val step = currentStep
        if (step != null) {
            step.addDoc(doc)
        } else {
            docs.add(doc)
        }
    }

    fun addAttachment(
        name: String,
        mediaType: String,
        path: String?,
        body: String?,
        encoding: String?,
        charset: String?,
        fileName: String?
    ) {
        val a = LinkedHashMap<String, Any?>()
        a["name"] = name
        a["mediaType"] = mediaType
        if (path != null) a["path"] = path
        if (body != null) a["body"] = body
        if (encoding != null) a["encoding"] = encoding
        if (charset != null) a["charset"] = charset
        if (fileName != null) a["fileName"] = fileName
        val step = currentStep
        if (step != null) {
            a["stepIndex"] = steps.indexOf(step)
            a["stepId"] = step.id
        }
        attachments.add(a)
    }

    fun getAttachments(): List<Map<String, Any?>> = attachments

    fun toStoryMeta(): StoryMeta = StoryMeta().apply {
        scenario = this@StoryContext.scenario
        steps = if (this@StoryContext.steps.isEmpty()) null else ArrayList(this@StoryContext.steps)
        tags = if (this@StoryContext.tags.isEmpty()) null else ArrayList(this@StoryContext.tags)
        tickets = if (this@StoryContext.tickets.isEmpty()) null else ArrayList(this@StoryContext.tickets)
        meta = if (this@StoryContext.meta.isEmpty()) null else LinkedHashMap(this@StoryContext.meta)
        docs = if (this@StoryContext.docs.isEmpty()) null else ArrayList(this@StoryContext.docs)
        sourceOrder = this@StoryContext.sourceOrder
    }

    fun startTimer(): Int {
        val token = timerCounter++
        val stepIndex = if (currentStep != null) steps.size - 1 else null
        val stepId = currentStep?.id
        activeTimers[token] = TimerEntry(System.nanoTime(), stepIndex, stepId)
        return token
    }

    fun endTimer(token: Int) {
        val entry = activeTimers[token] ?: return
        if (entry.consumed) return
        entry.consumed = true

        val durationMs = (System.nanoTime() - entry.startNanos) / 1_000_000.0

        var step: StoryStep? = null
        if (entry.stepId != null) {
            step = steps.find { it.id == entry.stepId }
        }
        if (step == null && entry.stepIndex != null && entry.stepIndex < steps.size) {
            step = steps[entry.stepIndex]
        }
        step?.durationMs = durationMs
    }

    companion object {
        private val ORDER_COUNTER = AtomicInteger(0)

        @JvmStatic
        internal fun resetOrderCounter() {
            ORDER_COUNTER.set(0)
        }
    }

    private fun isPrimary(keyword: String): Boolean =
        keyword == "Given" || keyword == "When" || keyword == "Then"
}
